### PR TITLE
Fix the Symfony 5.3 security deprecations

### PR DESCRIPTION
**Pull request by bytehead**
See https://github.com/contao/contao/pull/3215#issuecomment-891731721

### CODE REVIEW

**Comment by ausi** [comment]
You can increase all Symfony requirements from `^4.4 || ^5.2` to `^5.3`. This should make fixing this easier.

**Comment by ausi** [comment]
Also related: #2862

**Comment by bytehead** [comment]
All security related deprecations are gone except the ones we only can replace with the new authentication system (Contao 5.x):

`... is deprecated since Symfony 5.3, use the new authenticator system instead.`

File: core-bundle/src/Security/User/ContaoUserProvider.php [hunk 1]
```diff
@@ -50,18 +50,23 @@ public function __construct(ContaoFramework $framework, SessionInterface $sessio
     }
 
     public function loadUserByUsername($username): User
+    {
+        return $this->loadUserByIdentifier($username);
```
Comment (leofeyer) [inline core-bundle/src/Security/User/ContaoUserProvider.php:60]: Do we have to deprecate the `loadUserByUsername()` method as well?

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php [hunk 2]
```diff
@@ -55,13 +55,13 @@ protected function attemptAuthentication(Request $request): ?TokenInterface
         $currentToken = $this->tokenStorage->getToken();
 
         if ($currentToken instanceof TwoFactorTokenInterface) {
-            $authCode = $request->request->get('verify');
+            $authCode = $request->request->all()['verify'] ?? '';
```
Comment (leofeyer) [inline core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Why this change?

File: core-bundle/src/Security/User/ContaoUserProvider.php (hunk 1 above)
Comment (bytehead) [reply core-bundle/src/Security/User/ContaoUserProvider.php:60]: Fixed in ab36ce407.

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php (hunk 2 above)
Comment (bytehead) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/HttpFoundation/InputBag.php#L39

Comment (bytehead) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: But this (`non-string value`) is more a theoretical case during tests.

Comment (bytehead) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Fixed in cb47aa7d3.

**Review by leofeyer — DISMISSED** [review]

**Comment by leofeyer** [comment]
There are still some security-related deprecations when running the functional tests:

```
  2x: Since symfony/security-bundle 5.1: The "success_handler" at path "security.firewalls.contao_backend.logout" is deprecated, register a listener on the "Symfony\Component\Security\Http\Event\LogoutEvent" event instead.
    2x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration

  2x: Since symfony/security-bundle 5.1: The "success_handler" at path "security.firewalls.contao_frontend.logout" is deprecated, register a listener on the "Symfony\Component\Security\Http\Event\LogoutEvent" event instead.
    2x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration

  2x: Since symfony/security-bundle 5.3: The child node "encoders" at path "security" is deprecated, use "password_hashers" instead.
    2x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration
```

@bytehead Do you want to fix these here or in a separate PR?

**Comment by bytehead** [comment]
I'll fix them in this PR. 👍

**Comment by bytehead** [comment]
@leofeyer I've replaced the depreacted `LogoutSuccessHandler` with a corresponding listener. Somehow there is still some deprecations, but I don't know where they come from as I've adapted the security configuration for the functional tests as well.

```
   1x: Since symfony/security-http 5.1: The "Symfony\Component\Security\Http\Logout\DefaultLogoutSuccessHandler" class is deprecated, use "Symfony\Component\Security\Http\EventListener\DefaultLogoutListener" instead.
    1x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration

  1x: Since symfony/security-http 5.1: The "Symfony\Component\Security\Http\Logout\LogoutSuccessHandlerInterface" interface is deprecated, create a listener for the "Symfony\Component\Security\Http\Event\LogoutEvent" event instead.
    1x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration
```

**Comment by bytehead** [comment]
I've catched some more deprecations, the one mentioned [before](https://github.com/contao/contao/pull/3257#issuecomment-967229687) still exist. I have no clue what triggers them.

File: core-bundle/tests/Security/Logout/LogoutSuccessHandlerTest.php [hunk 3]
```diff
@@ -41,6 +47,8 @@ public function testRedirectsToAGivenUrl(): void
             ->willReturn(false)
         ;
 
+        $this->expectDeprecation('Since symfony/security-http 5.1: The "%s" class is deprecated, use "Symfony\Component\Security\Http\EventListener\DefaultLogoutListener" instead.');
```
Comment (ausi) [inline core-bundle/tests/Security/Logout/LogoutSuccessHandlerTest.php]: Deprecations that occur only once when the php file is autoloaded need to be handled in the *tests/PhpunitExtension/DeprecatedClasses.php* file, because the file could have already been autoloaded somewhere else and the deprecation would then not be triggered again.

**Comment by bytehead** [comment]
These won't be triggered while unit testing, only during the functional tests. I can't add them to 
`Contao\CoreBundle\Tests\PhpunitExtension\DeprecatedClasses`, as the unit test will fail then.

```
1x: Since symfony/security-http 5.1: The "Symfony\Component\Security\Http\Logout\DefaultLogoutSuccessHandler" class is deprecated, use "Symfony\Component\Security\Http\EventListener\DefaultLogoutListener" instead.
    1x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration

  1x: Since symfony/security-http 5.1: The "Symfony\Component\Security\Http\Logout\LogoutSuccessHandlerInterface" interface is deprecated, create a listener for the "Symfony\Component\Security\Http\Event\LogoutEvent" event instead.
    1x in RoutingMigrationTest::testShouldRun from Contao\CoreBundle\Tests\Functional\Migration
```

**Comment by bytehead** [comment]
And I don't know, if we should deprecate most of the security related classes as they won't exist in Contao 5 anymore?

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.
